packages/flutter_tools/lib/src/cache.dart

```diff
@@ -446,6 +477,25 @@ class Cache {
     return overrideUrl;
   }
 
+  /// The base for URLs that store Flutter engine artifacts in CIPD.
+  ///
+  /// For some platforms, such as Web and Fuchsia, CIPD artifacts are fetched
+  /// during the installation of the Flutter SDK, in addition to those fetched
+  /// from [storageBaseUrl].
+  ///
+  /// By default the base URL is https://chrome-infra-packages.appspot.com/dl.
+  /// However, if `FLUTTER_STORAGE_BASE_URL` environment variable is provided,
+  /// then the following value is used:
+  ///
+  ///     FLUTTER_STORAGE_BASE_URL/flutter_infra_release/cipd
```

Comment:
Is this manually maintained?

Reply:
In other words, if I add a new CIPD artifact to the tool, will it be present at https://storage.flutter-io.cn/flutter_infra_release/cipd? If not what is the process for getting it added? I would argue we should not allow further additions to this flow.

Reply:
I want to say yes, but I don't fully understand the question? Are the docs manually maintained? The code? Is there anything that can be automated?

Reply:
Ah, yes, it will be expected to be present at https://storage.flutter-io.cn/flutter_infra_release/cipd. I imagine @chenglu already has a process for identifying newly added artifacts and pushing them to the mirror. Whether we should be adding more CIPD artifacts is a little orthogonal. _If_ they are added, they will be downloaded from this URL. If not, then it doesn't matter.

Reply:
> I imagine @chenglu already has a process for identifying newly added artifacts and pushing them to the mirror.

Is this the case? Certainly they're not mirroring all of https://chrome-infra-packages.appspot.com/dl, right? If it's just the Flutter public ones, that might be feasible, but is that the case?

Reply:
> Whether we should be adding more CIPD artifacts is a little orthogonal. If they are added, they will be downloaded from this URL. If not, then it doesn't matter.

Agreed, if the mirroring is automated then that's good enough for me.

Reply:
Wouldn't you just mirror everything that's downloaded using `flutter precache --all-platforms`?

Reply:
IOW, you don't need all of CIPD, just the parts downloaded by the Flutter SDK, which can be enumerated using `flutter precache`. Actually, now that I think about it, `flutter precache --dry-run` would be a useful option. Just list what will be downloaded without downloading it. Then a mirror admin can get that list for the latest version of Flutter and mirror it.

Reply:
> Actually, now that I think about it, flutter precache --dry-run would be a useful option.

Agreed: https://github.com/flutter/flutter/issues/82383

> IOW, you don't need all of CIPD, just the parts downloaded by the Flutter SDK, which can be enumerated using flutter precache.

This is possible, although I'm pretty sure that's not how the mirroring works today. I also think this would be difficult to do without `flutter precache --dry-run`, as it would involve parsing the flutter_tools source to determine the remote paths.

Reply:
A workaround is to point `FLUTTER_STORAGE_BASE_URL` to `localhost` and use a local server that prints every URL fetched by Flutter and responds with an empty file :)

Reply:
[**https://chrome-infra-packages.appspot.com/dl**/flutter/web/canvaskit_bundle/+/1e8cK_8LOs0dz4lqd20LwTUYNqfu_4YL-dFG5yK1xXQC
](https://chrome-infra-packages.appspot.com/dl/flutter/web/canvaskit_bundle/+/1e8cK_8LOs0dz4lqd20LwTUYNqfu_4YL-dFG5yK1xXQC
)-->
[**https://storage.flutter-io.cn/flutter_infra_release/cipd**/flutter/web/canvaskit_bundle/+/1e8cK_8LOs0dz4lqd20LwTUYNqfu_4YL-dFG5yK1xXQC](https://storage.flutter-io.cn/flutter_infra_release/cipd/flutter/web/canvaskit_bundle/+/1e8cK_8LOs0dz4lqd20LwTUYNqfu_4YL-dFG5yK1xXQC)

This should be worked now.

Let me try to explain how the mirror works, which is essentially just a reverse host proxy that "whitelist" some of the gcs buckets like `dart-archive/ `,`download.flutter.io/` as well as `flutter_infra_release/`, so it can be accessed with the url `storage.flutter-io.cn/dart-archive/` , `storage.flutter-io.cn/download.flutter.io/`, `storage.flutter-io.cn/flutter_infra_release/`, etc. 

For the CIPD case, I pointed the path `flutter_infra_release/cipd` to `chrome-infra-packages.appspot.com/dl`, and the request will follow to get the redirected source (since the `chrome-infra-packages.appspot.com/dl` will 301 to `https://storage.googleapis.com/chrome-infra-packages/` with a looong gcs credentials). The Tsinghua University (tuna mirror) and Shanghai Jiao Tong University (sjtug mirror) are partially dependent on the flutter-io.cn mirror, so the changes we make will take effect on both of them as well.

But my personal suggestion is that it would be better if these resources could be copied directly from the internal into gcs://flutter_infra_release/cipd when the build is done, thanks!

Reply:
Cool, thanks for the explanation @chenglu

Reply:
> But my personal suggestion is that it would be better if these resources could be copied directly from the internal into gcs://flutter_infra_release/cipd when the build is done, thanks!

This will happen when we start building CanvasKit from sources as part of the engine build (@hterkelsen may be able to provide an ETA). The reason it is a CIPD package today is because we consume it as a binary that's prebuilt by the Skia team for us. It's temporary.